### PR TITLE
Rebuild the user directory and stats tables.

### PR DESCRIPTION
**Pull request by clokep**
Due to #14435, #14592, and #14639 I was curious if these tables were actually of date.

I ran the following on matrix.org to see if any room names in `room_stats_state` were different from `current_state_events`:

```sql
SELECT room_id, event_id, name AS stats_name, json::json->'content'->>'name' AS json_name
FROM current_state_events
INNER JOIN event_json USING (event_id, room_id)
INNER JOIN room_stats_state USING (room_id)
WHERE
    type = 'm.room.name' AND state_key = ''
    AND name != json::json->'content'->>'name';
```

(This returned results, I limited it to 3 since I didn't see any reason to get a full list.)

Thus, we should likely rebuild the tables which are affected by the state delta processing. To my knowledge this is just the user directory and user / room stats.

### CODE REVIEW

**Review by DMRobertson — APPROVED** [review]
Seems reasonable. Are there any other known stream cache change bugs we want to fix before landing this?

File: changelog.d/14643.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a long-standing bug where a device list update might not be sent to clients in certain circumstances.
```
[the remark this answers is not in the document]
Comment (clokep) [reply changelog.d/14643.bugfix:1]:
```suggestion
Fix a long-standing bug where the user directory and room/user stats might be out of sync.
```

@DMRobertson Does this sound better?

[the remark this answers is not in the document]
Comment (DMRobertson) [reply changelog.d/14643.bugfix:1]: Sounds good to me!

**Comment by DMRobertson** [comment]
I think we should have probably called this out in the upgrade notes; I will PR this retroactively.

**Pull request merged**